tee-worker/service/src/main.rs

```diff
@@ -787,27 +790,39 @@ fn send_extrinsic(
 fn subscribe_to_parentchain_new_headers<E: EnclaveBase + Sidechain>(
 	parentchain_handler: Arc<ParentchainHandler<ParentchainApi, E>>,
 	mut last_synced_header: Header,
-) -> Result<(), Error> {
-	let (sender, receiver) = channel();
-	//TODO: this should be implemented by parentchain_handler directly, and not via
-	// exposed parentchain_api. Blocked by https://github.com/scs/substrate-api-client/issues/267.
-	parentchain_handler
-		.parentchain_api()
-		.subscribe_finalized_heads(sender)
-		.map_err(Error::ApiClient)?;
-
+) {
 	loop {
-		let new_header: Header = match receiver.recv() {
-			Ok(header_str) => serde_json::from_str(&header_str).map_err(Error::Serialization),
-			Err(e) => Err(Error::ApiSubscriptionDisconnected(e)),
-		}?;
-
-		println!(
-			"[+] Received finalized header update ({}), syncing parent chain...",
-			new_header.number
-		);
-
-		last_synced_header = parentchain_handler.sync_parentchain(last_synced_header)?;
+		let (sender, receiver) = channel();
+		if let Err(e) = parentchain_handler.parentchain_api().subscribe_finalized_heads(sender) {
```

Comment:
It should work, thanks. A few Q:

- could it somehow spin off unlimited ws-connection (or background thread)? Is the graceful shutdown already handled?
- we only need some tests for it
- do you think it should be applied to upstream as well?

Reply:
Actually, this api calls the `substrate-client-api` crate, and in the substrate client api, it spawns a ws connect in a thread. there is no graceful shutdown.

I think it's no need to apply to the upstream, upstream should keep simple, just make reconnect when we use it.

Reply:
if we want to graceful shutdown, just break the `receiver` loop. when the sender sends data, it will fail, the tread will over with an error, then websocket connect will be shutdown.

Reply:
btw, the `graceful shutdown` is needed. Just created a new issue https://github.com/litentry/litentry-parachain/issues/1449.